pkg/gather/clusterconfig/clusterconfig.go

```diff
@@ -164,11 +169,40 @@ func GatherClusterOperators(i *Gatherer) func() ([]record.Record, []error) {
 					klog.V(2).Infof("Unable to find pods in namespace %s for failing operator %s", namespace, item.Name)
 					continue
 				}
-				for i := range pods.Items {
-					if isHealthyPod(&pods.Items[i], now) {
+				for j := range pods.Items {
+					pod := &pods.Items[j]
+					if isHealthyPod(pod, now) {
 						continue
 					}
-					records = append(records, record.Record{Name: fmt.Sprintf("config/pod/%s/%s", pods.Items[i].Namespace, pods.Items[i].Name), Item: PodAnonymizer{&pods.Items[i]}})
+					records = append(records, record.Record{Name: fmt.Sprintf("config/pod/%s/%s", pod.Namespace, pod.Name), Item: PodAnonymizer{pod}})
+
+					// Fetch previous and current pod logs
+					for _, previous := range []bool{true, false} {
+						for _, container := range pod.Spec.Containers {
+							currentOrPrevious := fmt.Sprintf("%s_current.log", container.Name)
+							if previous {
+								currentOrPrevious = fmt.Sprintf("%s_previous.log", container.Name)
+							}
+
+							req := i.coreClient.Pods(namespace).GetLogs(pod.Name, &corev1.PodLogOptions{Previous: previous, TailLines: &logTailLines, Container: container.Name})
+							readCloser, err := req.Stream()
+							if err != nil {
+								klog.V(2).Infof("Failed to fetch %s for %s pod in namespace %s for failing operator %s: %q", currentOrPrevious, pod.Name, namespace, item.Name, err)
+								continue
+							}
+
+							defer readCloser.Close()
+
+							buf := bytes.NewBufferString("")
```

Comment:
Log lines can be multimegabytes.  We had a bug recently where a single line was 15MB.  This means that you could hold potentially unlimited memory in the operator, causing a crash that would be unrecoverable.

You need to limit the maximum bytes the log can gather (MaxBytes field) and reuse the buffer instead of initializing it for every pod.

You also need to consider what the maximum number of pods could be read this way and limit the size based on number of total containers you are gathering logs for.  If you are gathering logs for 100 failing containers, 500k might be acceptable.  If this is a huge cluster (5k nodes) and you have a crashlooping daemonset, it might be 20k containers, in which case 500k is way too big.  You MUST bound total memory consumption in this operator to some reasonable limit (50mb is about what i might allow).

The real solution here is that you are misuing Record - you want a record type that is backed by a file (and you should be streaming these logs to a temporary file location and then cleaning them up after they are copied to the tar, or moving them from temp dir to final location at the time record is committed).  Record is not supposed to be used for unbounded large objects.

Reply:
>You need to limit the maximum bytes the log can gather (MaxBytes field) and reuse the buffer instead of initializing it for every pod.

Implemented in https://github.com/openshift/insights-operator/pull/126

>The real solution here is that you are misuing Record - you want a record type that is backed by a file

Agree, we're storing arbitrary user certificates from config maps too, so a file-backed storage might be useful in other places. @martinkunc do you think this is worth a jira story?

Reply:
I would probably say any one gatherer should never exceed 1-5mb in memory at any one time, and so for instance if you read multiple things we need records that can be lazily put on temp disk and then moved to the upload / record location on demand